### DISCUSSION TITLE
Reliable way to tell rosetta virtualized vm from inside the guest

### DISCUSSION THREAD

**Discussion by goodwillcoding** [lima-vm/lima#2340] [Q&A] [answered]
Hi folks,

What's a reliable way to the guest is running a virtualized to know it supports. 
Using https://raw.githubusercontent.com/lima-vm/lima/master/examples/experimental/vz.yaml where rosetta is enabled on M* macOS
```
$ limactl start --name test1 vz.yaml
```
And uname is not helping
```
$ limactl shell ubuntu uname -a
Linux lima-ubuntu 6.8.0-31-generic #31-Ubuntu SMP PREEMPT_DYNAMIC Sat Apr 20 02:32:42 UTC 2024 aarch64 aarch64 aarch64 GNU/Linux
```
What am I missing?

**Comment by afbjorklund** [comment, marked as answer]
As far as I know, Rosetta is registered using binfmt_misc - so it should be in `/proc/sys/fs/binfmt_misc`

Note that Rosetta is "only" a binary translator (like qemu-user) and not a hypervisor (like qemu-system)

**Reply by afbjorklund** [reply]
To guess the hypervisor (in a distribution using systemd), you can use: `lima systemd-detect-virt`

For containers, it only works if they tell it. And for containerd, it will simply say `container-other`

**Reply by afbjorklund** [reply]
Actually it _will_ try some heuristics to find Docker and Podman, based on environment files:

https://github.com/systemd/systemd/blob/v255/src/basic/virt.c#L631 (others use `$container`)

**Reply by goodwillcoding** [reply]
> Note that Rosetta is "only" a binary translator (like qemu-user) and not a hypervisor (like qemu-system)
Right but something is confusing ... is not it translating arm <-> x86 ... so why would the arch for the guest be arm, is the point to run x86 binaries ... which in turn are able to virtualize infra for x86 ...
```
What am I missing?

**Reply by goodwillcoding** [reply]
I guess it's limitation of rosetta based on https://kb.parallels.com/129871 ... so it give arm ability to run x86 while staying arm ... I guess that's a lot of scripts to change that use `uname` (or fake uname)

**Reply by afbjorklund** [reply]
Just meant that "rosetta" is not the virtualization method, it is called "apple" (Apple's Virtualization.framework, or "VZ")

As you mention, Rosetta is used to run amd64 binaries on arm64. But it can do that both on the host, or in virtualization.

**Reply by afbjorklund** [reply]
Ah, "Rosetta.Enabled" actually refers to [VZLinuxRosettaDirectoryShare](https://developer.apple.com/documentation/virtualization/vzlinuxrosettadirectoryshare) (the file system used)

The actual setting is "Rosetta.BinFmt", that enables the binary translation... Good to know. :-)

**Comment by afbjorklund** [comment]
The file system is mounted under `/mnt/lima-rosetta`. The handler is `/proc/sys/fs/binfmt_misc/rosetta`.

They should correspond to the Rosetta.Enabled and Rosetta.Binfmt settings, in the instance configuration

```yaml
rosetta:
  # Enable Rosetta for Linux (EXPERIMENTAL).
  # Hint: try `softwareupdate --install-rosetta` if Lima gets stuck at `Installing rosetta...`
  # 🟢 Builtin default: false
  enabled: null
  # Register rosetta to /proc/sys/fs/binfmt_misc
  # 🟢 Builtin default: false
  binfmt: null
```

https://lima-vm.io/docs/config/multi-arch/#fast-mode-2-rosetta-intel-containers-on-arm-vm-on-arm-hostfast-mode-2
